prow/tide/history/history.go

```diff
@@ -83,22 +83,53 @@ func readHistory(maxRecordsPerKey int, opener io.Opener, path string) (map[strin
 }
 
 func writeHistory(opener io.Opener, path string, hist map[string][]*Record) error {
-	writer, err := opener.Writer(context.Background(), path)
-	if err != nil {
-		return fmt.Errorf("open: %v", err)
-	}
 	b, err := json.Marshal(hist)
 	if err != nil {
 		return fmt.Errorf("marshal: %v", err)
 	}
-	if _, err := fmt.Fprint(writer, string(b)); err != nil {
+	data := string(b)
+	for i := 0; i < 5; i++ {
+		timedOut, err := writeHistoryWithTimeout(opener, path, data)
+		if timedOut && i != 4 {
+			logrus.WithError(err).Warn("Timed out flushing history, retrying.")
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		break
+	}
+	return nil
+}
+
+// writeHistoryWithTimeout attempts to write the history with a timeout, returning
+// whether the timeout was reached and any errors that occurred
+func writeHistoryWithTimeout(opener io.Opener, path string, hist string) (bool, error) {
+	// a write's duration will scale with the volume of data to write but large
+	// data sets can finish in about 500ms; a timeout of 30s should not evict
+	// well-behaved writes
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	// we need to keep track of whether the context timed out or not
+	var timedOut bool
+	go func(done <-chan struct{}) {
+		<-done
+		timedOut = true
```

Comment:
I think this may be racy?

Reply:
Would love ideas of how else to do this.

Reply:
I don't think we need all this extra complexity -- go routines, a whole second method.

Let's just make `writeHistory()` use a 30s timeout. If it fails we'll try again in a few minutes.